### ISSUE TITLE
linter: `jsPlugins` — Multi-threaded execution silently drops warnings from plugins using `synckit`

### ISSUE THREAD

**Issue by taichi221228** [oxc-project/oxc#19986]
### What version of Oxlint are you using?

1.50.0

### What command did you run?

oxlint --type-aware

### What does your `.oxlintrc.json` (or `oxlint.config.ts`) config file look like?

```jsonc
import { defineConfig } from "oxlint";
// @ts-expect-error @pandacss/eslint-plugin has no type declarations
import pandacss from "@pandacss/eslint-plugin";

export default defineConfig({
  jsPlugins: [{ name: "@pandacss", specifier: "@pandacss/eslint-plugin" }],
  rules: {
    ...pandacss.configs.recommended.rules,
    "@pandacss/no-escape-hatch": "warn",
    "@pandacss/no-important": "warn",
    "@pandacss/prefer-longhand-properties": "warn",
    "@pandacss/prefer-atomic-properties": "warn",
  },
});
```

### What happened?

When using `jsPlugins` with a plugin that internally uses `synckit` (such as
`@pandacss/eslint-plugin`), oxlint silently drops all warnings from that plugin
when more than one thread is used (the default).

No diagnostic is emitted — the output simply shows `Found 0 warnings and 0 errors`.
The plugin's error handler does call `console.error` internally, but that output either
does not appear in the terminal or is suppressed by oxlint's worker thread environment,
making the issue very difficult to diagnose.

### Steps to Reproduce

1. Install `@pandacss/eslint-plugin` and configure it via `jsPlugins`:

   ```ts
   // oxlint.config.ts
   import { defineConfig } from "oxlint";
   import pandacss from "@pandacss/eslint-plugin";

   export default defineConfig({
     jsPlugins: [{ name: "@pandacss", specifier: "@pandacss/eslint-plugin" }],
     rules: {
       ...pandacss.configs.recommended.rules,
       "@pandacss/prefer-atomic-properties": "warn",
     },
   });
   ```

2. Create a source file with a violation:

   ```tsx
   // src/App.tsx
   import { css } from "./styled-system/css";

   export function App() {
     return <div className={css({ background: "red", padding: "4" })} />;
   }
   ```

3. Run `oxlint` (default: uses multiple threads):

   ```
   Found 0 warnings and 0 errors.
   ```

   Expected warnings are silently missing.

4. Run `oxlint --threads=1`:

   ```
   ! @pandacss(prefer-atomic-properties): Use atomic properties instead of `background`. ...
   ! @pandacss(prefer-atomic-properties): Use atomic properties instead of `padding`. ...
   Found 2 warnings and 0 errors.
   ```

   Warnings appear correctly.

### Root Cause (Hypothesis)

`@pandacss/eslint-plugin` uses `synckit` internally. `synckit` blocks the calling
thread via `Atomics.wait()` until a dedicated worker thread responds. When
synchronization fails, the plugin's error handler catches the exception, logs it to
stderr (which appears to be suppressed in oxlint's worker environment), and returns
`undefined` — causing the rule to emit no diagnostic.

The exact failure mechanism in oxlint's multi-threaded environment has not been
confirmed. The `--threads=1` workaround avoids the issue by ensuring only one
thread calls the plugin at a time.

### Expected Behavior

Warnings from `@pandacss/eslint-plugin` are reported correctly regardless of the
number of threads, as they are with `--threads=1`.

### Actual Behavior

Warnings are silently dropped. No diagnostic output or error message appears, so
the failure is completely invisible to the user.

### Environment

|                         |                             |
| ----------------------- | --------------------------- |
| oxlint                  | 1.50.0                      |
| @pandacss/eslint-plugin | 0.3.1                       |
| Node.js                 | v25.4.0                     |
| OS                      | macOS Darwin 25.3.0 (arm64) |

### Workaround

Add `--threads=1` to the oxlint command:

```json
{
  "scripts": {
    "lint": "oxlint --threads=1"
  }
}
```

**Comment by camc314** [comment]
https://github.com/chakra-ui/eslint-plugin-panda/blob/4a9fcae055d712dc37cc11b5003e5af0b0f03c43/plugin/src/utils/helpers.ts#L81

**Comment by camc314** [comment]
In oxlint, file order is non-deterministic, as linting files in different orders should produce the same result. Given this, we make no guarantees about the order in which files should be linted in.

`eslint-plugin-panda` makes use of a cache in the form of a `WeakMap<RuleContext<any, any>, ImportResult[]>`.

This caches the results of calculating imports within files, reducing work across rules.

Oxlint reuses the same context object across files to reduce GC pressure and help improve performance.

This means that the same key is being used for the weakmap cache, so we get stale import results depending on the order that files are linted in.

cc @overlookmotel

**Comment by overlookmotel** [comment]
There's 2 different things going on here:

### Missing stderr from worker threads

It's odd that behavior is different depending on `--threads` setting. `--threads` only controls how many threads are used for linting *on Rust side*. Regardless of `--threads`, all JS plugins currently run on a single JS thread.

###  `context` object

2nd problem, as @camc314 said, is that Oxlint differs from ESLint in that there's a single `context` object which is reused for every file. This is more performant.

I'd be loathe to give up the perf gain to support a fairly niche use case (running async work synchronously via worker threads and `Atomics`).

I think it'd be preferable to make a PR on `@pandacss/eslint-plugin` to make it compatible with Oxlint. Hopefully it'd be accepted. I'm not 100% sure what that PR would look like, but we can look into it.

**Comment by camc314** [comment]
> Missing stderr from worker threads

> It's odd that behavior is different depending on --threads setting. --threads only controls how many threads are used for linting on Rust side. Regardless of --threads, all JS plugins currently run on a single JS thread.

I don't htink this is true - logging from worker threads worked fine in my testing.

> @pandacss/eslint-plugin uses synckit internally. synckit blocks the calling thread via Atomics.wait() until a dedicated worker thread responds. When synchronization fails, the plugin's error handler catches the exception, logs it to stderr (which appears to be suppressed in oxlint's worker environment), and returns undefined — causing the rule to emit no diagnostic.

Note that this root cause hypothis is incorrect

**Comment by overlookmotel** [comment]
Having looked at `@pandacss/eslint-plugin`, I think it's caching strategy is not optimal in any case. It's caching data which is per-file and could be shared across all rules, but that's not happening with a `WeakMap` keyed by `Context` - in ESLint each rule gets a separate `Context` object, so their cache is per-file-per-rule, not per-file.

I've made a PR which would fix that, and in the process should also make the plugin work with Oxlint: https://github.com/chakra-ui/eslint-plugin-panda/pull/308

**Comment by overlookmotel** [comment]
Note: ESLint's `verifyAndFix()` (used in `--fix` mode) calls `verify()` synchronously in a loop up to 10 times for the same file.

The cache implementation in my PR https://github.com/chakra-ui/eslint-plugin-panda/pull/308 only resets the cache on next micro-tick, so it would get stale data on the 2nd and later passes.

That fine in the case of `@pandacss/eslint-plugin`, because none of the rules provide auto-fixes, only suggestions, so they wouldn't cause `verifyAndFix()` to be called. No problem.

But this solution isn't completely generalizable - plugins which do provide auto-fixes *and* rely on caching of this kind would need a different solution.

Here's an alternative which doesn't rely on `queueMicrotask`, and so avoids this problem: https://github.com/chakra-ui/eslint-plugin-panda/pull/309

But that's a bit less ergonomic - you have to wrap every rule definition with a `createRule(...)` call to add the "how many rules currently running" counter. It's probably still more performant than using `WeakMap`s though.

ESLint's API unfortunately doesn't offer a nice way to share state between rules on a per-file basis.

**Comment by overlookmotel** [comment]
https://github.com/chakra-ui/eslint-plugin-panda/pull/309 was merged.

I think we should offer a more ergonomic way to implement data-sharing between rules in our alternative `createOnce` API.

**Comment by ghost** [comment]
> all JS plugins currently run on a single JS thread.

Is there a plan to make it run in multi-thread?

**Comment by camc314** [comment]
> 
> 
> > all JS plugins currently run on a single JS thread.
> 
> Is there a plan to make it run in multi-thread?

yes

**Comment by ghost** [comment]
> > > all JS plugins currently run on a single JS thread.
> > 
> > 
> > Is there a plan to make it run in multi-thread?
> 
> yes

When? Sorry I can't find it in the milestone.

**Comment by camc314** [comment]
> > > > all JS plugins currently run on a single JS thread.
> > > 
> > > 
> > > Is there a plan to make it run in multi-thread?
> > 
> > 
> > yes
> 
> When? Sorry I can't find it in the milestone.

It's not explicitly mentioned, but it would be part of the _"Improve performance"_ part of https://github.com/oxc-project/oxc/issues/19918
